### PR TITLE
feat(eval): per-scene clip tooling, plus review fixes for the recovery evals

### PR DESCRIPTION
**Pull request by danielsanchezaran**
# What

Follow-up to [#384](https://github.com/tier4/Diffusion-Planner/pull/384). Two new tools that turn a row of the recovery table back into something you can watch, plus the fixes from #384's review round.

| | |
|---|---|
| `rlvr/autoresearch/tools/find_disagreeing_starts.py` | **new** — which starts one arm recovers and another loses |
| `rlvr/autoresearch/tools/render_recovery_clip.py` | **new** — render one of those starts as a side-by-side clip |
| `rlvr/autoresearch/tools/eval_recovery_route.py`, `eval_incut_route.py` | review fixes (see below) |

**This PR contains #384's commit as well** — its branch could not be updated (`GH013 … Cannot update this protected ref` on both force-push and fast-forward), so the follow-up work is delivered here on a fresh branch. Reviewing this one supersedes #384.

# Why the clip tooling belongs in the repo

The evals produce rates. The evidence a reader actually looks at is a clip of one start where two arms did different things — and that step was still being done by scratch scripts. Two consequences, both of which have already happened:

- a write-up could show a clip nobody could regenerate;
- the caption numbers came from a different code path than the tables, so they could disagree with the row they illustrate and nobody would know.

Both tools import their rollout settings, perturbation hook and scoring path from `eval_recovery_route` rather than reimplementing them, so a clip is generated under the same settings as the measurement **by construction**. Verified end to end: for start 1258 the clip reports `settle=0.1667` / `LOST`, which is exactly what the eval rows say for that start.

`find_disagreeing_starts` keys rows by `(route, start, offset)`. An eval spans more than one recorded drive and start indices collide between them, so a `(start, offset)` key silently mixes two different places on the map; the scratch version worked around this by globbing one route's shards only.

# Review fixes from #384

- **[BLOCKER] Raw weights were scored silently.** A training milestone carries both `model` (raw optimizer iterates) and `ema_state_dict`, and `simulate.load_model` takes `model`. Pointing either tool at `epochNNNN/best_model.pth` therefore scored the raw iterates with no warning — a plausible but wrong verdict. `require_deployable_checkpoint` now refuses such a checkpoint and prints the conversion snippet.
- **Rollouts vanished from the denominator.** 1–9-step rollouts were dropped by a bare `continue`. Still excluded (their settle window would be the perturbation itself), now counted and reported as `skipped_short_rollouts`.
- **`route_distance_m` returned `inf`** when a frame had neither `route_lanes` nor `lanes`, marking every rollout lost — "drove off the map" rather than "the input has no map". Raises now.
- **`json.dumps` emitted bare `NaN`**, which no strict JSON reader accepts, in files whose purpose is to be aggregated across shards. Non-finite floats are written as `null`.
- **In-cut `frac_incut` was 0.0 when nothing was scored**, reading as "never cuts corners" off an all-NaN array. It and its siblings return `None`; `_turn_side` keeps its keys so an aggregator sees `None`, not `KeyError`.
- Rows now record `terminated`; each run writes `<out_json>.meta.json` with argv, checkpoints and rollout settings; the 5 m coverage radius is one named constant shared by both tools; recovery's argparse defaults are the protocol the published numbers were produced with.

Docstring corrections: the draw-pool note no longer claims the failure is silent (`collect_usage` raises on it), and `speed` is documented as unable to show reversing — the tracker reports a non-negative magnitude, so measure it from consecutive `ego` positions in `rollout.jsonl`.

Not taken: per-step signed curvature for the in-cut window. It would change what the metric measures, and the job here is to land the measurement as it was made; the limitation is in the module docstring instead.

# A bug the end-to-end test caught

`drive_rollout` never returned `render_segment`'s outcome, so `terminated` — the field added in response to review, the one that says whether a rollout reached the goal or merely ran out of budget — was **`null` on every row**. Lint, complexity and unit tests all passed; only running the eval and reading the output found it. Fixed, re-run, now `terminated='max_steps'`.

# Verification

- Both evals reproduce the campaign scripts they replace with **worst delta exactly 0** across `recovered`/`lost`, every summary statistic, and every per-step array including realized poses (recovery: 6 rollouts; in-cut: 5 curve rollouts).
- Clip tooling exercised end to end on real checkpoints: the picker found 5 of 42 shared rollouts where one arm recovers and the other is lost; the renderer produced both panes plus the `hstack` pair, with a report matching the eval rows.
- `rlvr/test_incut_geometry.py`: 6 tests pinning the sign convention on synthetic arcs — pass.
- `ruff check` + `ruff format --check` clean on all six files.
- Complexity: every block A or B (worst B(10)); average **A (4.05)**.

# Usage

```
python -m rlvr.autoresearch.tools.find_disagreeing_starts \
    --rows $CAMP/recovery/*.json --win TREATED --lose INCUMBENT --limit 6

python -m rlvr.autoresearch.tools.render_recovery_clip \
    --models incumbent:<a.pth> treated:<b.pth> --npz_root <route dir> \
    --start <n> --offset 1.0 --drop_objects --hstack --out_dir $CAMP/clips
```

Written up, with the measured output of both commands, in [How to A/B a training-recipe change](https://tier4.atlassian.net/wiki/spaces/ATT/pages/5531206307) §5c.

### CODE REVIEW

File: rlvr/autoresearch/tools/render_recovery_clip.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    ap.add_argument("--device", default="cuda")
+    ap.add_argument("--hstack", action="store_true", help="also emit the side-by-side clip")
+    ap.add_argument("--out_dir", required=True)
+    args = ap.parse_args()
+
+    reproducer_rollout._ego_state_from_frame = _offset_ego_state
+    reproducer_rollout._draw_step = _scoring_draw_step
+
+    routes = enumerate_routes(Path(args.npz_root))
+    key = sorted(routes)[0]
```
Comment (danielsanchezaran) [inline rlvr/autoresearch/tools/render_recovery_clip.py]: `find_disagreeing_starts` deliberately keys and prints each pick by `(route, start, offset)` because start indices collide across routes, but this renderer has no `--route` option and unconditionally selects the lexicographically first route. I reproduced it with two routes: a pick belonging to route B could only render route A. That can produce a plausible clip whose caption numbers came from a different recorded drive. Please require/select the route key emitted by the picker, fail if it is absent, include it in the output stem/report, and add a two-route regression test.

**Review by danielsanchezaran — COMMENTED** [review]
The fixes carried over from #384 correctly reject raw/EMA training milestones, account for short rollouts, reject missing route geometry, serialize non-finite values as strict JSON, and preserve rollout termination outcomes. The branch also merges cleanly with current `tier4-main`.

I found one evidence-integrity blocker in the new clip workflow: the disagreement picker identifies a specific route, but the renderer cannot accept that route and always opens the first one. A selected disagreement from any other route therefore renders a different scene at the same numeric start index. Please wire route identity through the renderer before relying on these clips.

The six geometry tests and six relevant reproducer tests pass. Ruff, formatting, Python compilation, and diff checks also pass. There are no GitHub checks attached to this head. GitHub does not allow me to request changes on my own PR, so I am leaving the blocker as a comment review.

**Comment by danielsanchezaran** [comment]
Fixed in `e7cb5be73` — you were right, and the failure mode was worse than a wrong picture: it broke the exact property this tooling was sold on, that a caption cannot disagree with the row it illustrates.

`select_route()` is now a pure function that honours `--route`, accepts a single-route root without one, and **refuses** a multi-route root rather than choosing for the caller. The route is threaded into the run, the output stem, the PNG title and the report (which now records route/start/offset alongside the per-arm verdicts), so two drives' clips cannot collide on disk either.

`find_disagreeing_starts` now prints the pick in the renderer's own flag form — `--route <key> --start <n> --offset <v>` — so a pick is pasted with its route or not at all:

```
5 of 42 shared rollouts have frenet recovering and quintic lost
  --route 15-32-12_00000000 --start 1242 --offset -1.0   # frenet 0.000 | quintic LOST
```

Regression test as requested: `rlvr/test_clip_route_selection.py`, four cases, including the two-route case that previously returned the lexicographically first route. Also verified by hand that a bogus `--route` now exits with the available keys instead of rendering something plausible.

Checks on this head: 10 tests pass (6 geometry + 4 route), ruff and ruff-format clean on all six files, radon average A(4.05) with nothing above B.

Worth noting how it hid: I only ever ran the picker and renderer against a single-route directory, where "first route" happens to be the right route.

**Review by danielsanchezaran — COMMENTED** [review]
Follow-up on `e7cb5be73`: the route-selection blocker is resolved. Multi-route roots now require an explicit route key, unknown keys fail loudly, the disagreement picker prints the renderer-ready `--route` form, and route identity is carried through clip filenames, captions, and reports. The added two-route regression tests cover the original wrong-route failure. I found no new correctness issues.

The 16 relevant geometry, route-selection, and reproducer tests pass on both the exact head and GitHub's current test merge. Ruff, formatting, Python compilation, and `git diff --check` pass. There are no GitHub checks attached to this head. GitHub does not allow me to approve my own PR, so this is submitted as a comment review, but the current head is clear from this review.

**Comment by Owen-Liuyuxuan** [comment]
@danielsanchezaran I found two concrete correctness issues that look worth fixing before relying on the reported recovery rows and generated clips:

1. **A rollout that ends off-route can still be classified as recovered.**

   In `summarize_rollout`, `settle_near = usage[-10:][near[-10:]]` removes off-route samples from the final window. `lost` then uses `not near[-10:].any()`, so only one near-route sample anywhere in the last ten is enough to avoid that branch.

   A concrete 80-step trace is 71 near-route steps followed by 9 off-route steps, with usage 0 on the one near sample in the final window:

   - `off_frac = 9 / 80 = 0.1125`, so `off_frac > 0.5` is false;
   - `near[-10:].any()` is true because the first sample in that window is near;
   - `settle_near == [0]`, so its mean is `0 <= 0.5`.

   The exact function therefore returns `lost=False, recovered=True`, even though the final pose and final nine steps are off-route. That contradicts the adjacent comment that recovered means the rollout "ended ... back on the route." At minimum, recovery appears to need `near[-1]`; alternatively the protocol could require a defined near-route fraction across the complete settle window.

2. **Rerendering the same clip identity can encode stale PNGs from an earlier run.**

   `render_one` reuses `out_root / f"{stem}_{label}"` with `exist_ok=True`, without clearing or uniquely versioning it. `encode_webm` subsequently gives ffmpeg the glob `png_dir / "*.png"`.

   For example, a first run writes `00000.png` through `00079.png`. A second run with the same route/start/offset/world/label that stops after 50 frames overwrites `00000.png` through `00049.png`, but leaves the old `00050.png` through `00079.png`. ffmpeg then encodes all 80 files, while the JSON verdict was computed only from the new in-memory 50-step score buffers. The resulting video can therefore disagree with its report.

   Please render into a fresh run-specific directory, or safely reject/clean an existing frame directory before starting. A regression test should prepopulate higher-numbered PNGs, perform a shorter rerender, and verify that only current-run frames reach the encoder.

**Comment by danielsanchezaran** [comment]
@Owen-Liuyuxuan both confirmed and fixed in `b01259674`, with regression tests.

**1. Off-route ending classified as recovered — confirmed with your exact trace.** 71 near / 9 off with usage 0 on the lone near sample returned `lost=False, recovered=True` on the previous scorer. The cause is the one you named: `lost` used `near[-10:].any()`, and `settle_near = usage[-10:][near[-10:]]` then averaged over whatever survived that filter — a single sample in this case.

I went with a three-way verdict rather than picking one of your two alternatives, because they answer different questions and forcing a binary is what created the hole:

- `lost` = ended off the route (`not near[-1]`), or >50% of the rollout off it — the literal reading
- `recovered` = not lost **and** the whole settle window on the route **and** its mean usage ≤ 0.5. Off-route steps are never filtered out of the window any more.
- `unsettled` = neither: ended on-route but was still swerving through the coverage radius during the last second

`summarize_model` now reports `unsettled_rate` alongside the other two, and `SETTLE_WINDOW` is a named constant next to the coverage radius.

Measured impact, re-scored from saved per-step traces of five ep200 checkpoints (~1,113 rollouts each): recovered% drops 0.5–0.8 pts and lost% rises 2–5 pts on every arm — 27–58 rollouts per arm had ended off-route but counted as recovered. The swerving-but-ended-on-route rollouts are 7–11% of each arm; under the previous scorer they landed on either side depending on where the single near sample fell. Every arm shifts by a similar amount, so relative verdicts between arms scored together survive; absolute figures quoted from the old scorer do not, and I'll re-state the ones in circulation.

**2. Stale frames on rerender — confirmed and fixed.** `fresh_frame_dir()` empties the frame directory of PNGs before every render, so only the current run's frames can reach the `*.png` glob. The regression test does what you described: prepopulates 80 frames, rerenders 50, asserts exactly 50 current frames remain.

Tests in `rlvr/test_recovery_verdict.py`: your trace → lost; swerve-then-back → unsettled (not lost, not recovered); off-route steps cannot be dropped from the settle mean; a clean settle → recovered; and the frame-directory case. The verdict test fails against the previous scorer and passes with this one.

**Review by SakodaShintaro — APPROVED** [review]
